samples/can/src/main.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2018 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <misc/printk.h>
+#include <device.h>
+#include <can.h>
+#include <gpio.h>
+
+#define CAN_DEV "CAN_1"
+#define GPIO_LED_DEV "GPIOC"
+#define GPIO_BUTTON_DEV "GPIOA"
+#define PIN_USER_BUTTON 0
+#define PIN_LED_1 6
+#define PIN_LED_2 7
+#define PIN_LED_3 8
+#define PIN_LED_4 9
+#define TX_THREAD_STACK_SIZE 512
+#define LED_THREAD_STACK_SIZE 512
+#define RX_STR_THREAD_STACK_SIZE 512
+#define TX_THREAD_PRIORITY 2
+#define LED_MSG_ID (0x10)
+#define BUTTON_MSG_ID (0x01)
+#define STR_MSG_ID (0x12345)
+#define LOOPBACK 1
+
+K_THREAD_STACK_DEFINE(tx_thread_stack, TX_THREAD_STACK_SIZE);
+K_THREAD_STACK_DEFINE(led_thread_stack, LED_THREAD_STACK_SIZE);
+K_THREAD_STACK_DEFINE(rx_str_thread_stack, RX_STR_THREAD_STACK_SIZE);
+struct k_thread tx_thread_data;
+struct k_thread led_thread_data;
+struct k_thread rx_str_thread_data;
+struct k_sem tx_sem;
+static struct gpio_callback gpio_cb;
+CAN_DEFINE_MSGQ(led_msgq, 2);
+CAN_DEFINE_MSGQ(str_msgq, 5);
+
+void tx_irq_callback(u32_t error_flags)
+{
+	printk("Callback! error-code: %d\n", error_flags);
+}
+
+void button_callback(struct device *port,
+				   struct gpio_callback *cb, u32_t pins)
+{
+	k_sem_give(&tx_sem);
+}
+
+void send_string(char* string, struct device *can_dev)
+{
+	struct can_msg msg;
+	msg.ext_id = STR_MSG_ID;
+	msg.id_type = can_extended_identifier;
+	msg.dlc = 0;
+	msg.rtr = can_dataframe;
+	for(int str_len = strlen(string); str_len;)
+	{
+		msg.dlc = str_len >= 8 ? 8 : str_len;
+		str_len -= msg.dlc;
+		memcpy(msg.data, string, msg.dlc);
+		string += msg.dlc;
+		can_send(can_dev, &msg, 10, tx_irq_callback);
+	}
+	
+}
+
+void tx_thread(void *can_dev_param, void *unused2, void *unused3)
+{
+	struct can_msg msg;
+	struct can_msg msg_button_cnt;
+	struct device *can_dev = can_dev_param;
+	static uint8_t counter = 0;
+	static u16_t button_press_cnt = 0;
+
+	msg.std_id = LED_MSG_ID;
+	msg.id_type = can_standard_identifier;
+	msg.dlc = 1;
+	msg.rtr = can_dataframe;
+	msg.data[0] = 0;
+
+	msg_button_cnt.std_id = BUTTON_MSG_ID;
+	msg_button_cnt.id_type = can_standard_identifier;
+	msg_button_cnt.dlc = 2;
+	msg_button_cnt.rtr = can_dataframe;
+	msg_button_cnt.data[0] = 0;
+	msg_button_cnt.data[1] = 0;
+
+	printk("TX thread is running.\n");
+	while (1)
+	{
+		k_sem_take(&tx_sem, K_FOREVER);
+		button_press_cnt++;
+		counter = counter > 3 ? 0 : counter + 1;
+		printk("Button pressed! Send message %u\n", counter);
+		msg.data[0] = counter;
+		msg_button_cnt.data[0] = button_press_cnt & 0xFF;
+		msg_button_cnt.data[1] = (button_press_cnt >> 8) & 0xFF;
+		can_send(can_dev, &msg, 10, tx_irq_callback);
+		can_send(can_dev, &msg_button_cnt, 10, NULL);
+		if(counter == 4)
+		{
+			send_string("Counter is at 4. Reset\n", can_dev);
+		}
+	}
+}
+
+void rx_str_thread(void *msgq, void *can_dev_param, void *unused)
+{
+	struct can_msg msg;
+	{
```

Comment:
having {} here looks unnecessary.

Reply:
This is to create a scope for the filter. I wanted to show that the filter struct can be discarded after the attach function.

Reply:
well, document can_attach function in the api about that behavior then (if not already).

This kind of lone block is a bit ugly